### DISCUSSION TITLE
Is there a way to trigger the build of a specific (not known to be active) configuration of a CDT project?

### DISCUSSION THREAD

**Discussion by jantje** [eclipse-cdt/cdt#355] [Q&A] [answered]
Hi
The standard eclipse framework supports configurations but as far as I can see it is not fully implemented.
Eclipse allows to reference projects (but not their configurations)
CDT allows to reference projects and their configurations.
But as the eclipse build command is at the project level (and not the configuration level) how can one trigger a build of a specific configuration (like the CDT gui requires to be possible)?
I noticed in the MBS commonBuilder the code assumes that the referenced projects are MBS so it can directly call the commonBuilder build. 
However the CDT gui seems to allow to reference any CDT type project. So it seems to me the assumption made by the MBS code is wrong.

This means that referencing configurations of CDT(, but non MBS) projects from MBS projects may not trigger a build of the specified configuration (but rather trigger a build of the active configuration).
I expected something like `ICConfigurationDescription.build(int buildType, IProgressMonitor monitor)` but I fail to find something like that.
Therefore I'm asking:
Is there a way to trigger the build of a specific (not known to be active) configuration of a CDT project?
Thanks
Jantje

**Comment by jonahgraham** [comment, marked as answer]
There is, have a look at how the headless builder does it, entry point is https://github.com/eclipse-cdt/cdt/blob/main/build/org.eclipse.cdt.managedbuilder.headlessbuilderapp/src/org/eclipse/cdt/managedbuilder/internal/headlessbuilderapp/HeadlessBuilderWrapper.java

**Reply by jantje** [reply]
Hi Jonah
Thanks for the reply.
I investigated the headless way and if I'm not mistaken the headless uses the arguments of
`protected abstract IProject[] build(int kind, Map<String, String> args, IProgressMonitor monitor) throws CoreException;`
It looks to me that CDT internal build processing does not handle the args.
So it is up to the final implementation of the build to unpack the arguments (and find the provided configuration).
Checking the builders I only found 2 builders that actually process the args parameter. Those are `CommonBuilder `and `MakeBuilder`.
These are the other builders I checked:
```
AutotoolsConfigurationBuilder
GeneratedMakefileBuilder
GNUMakeFileChecker
ScannerConfigBuilder/org.eclipse.cdt.make.core/src/org/eclipse/cdt/make/core/scannerconfig/ScannerConfigBuilder.java@Deprecated
/org.eclipse.cdt.managedbuilder.core/src/org/eclipse/cdt/build/core/scannerconfig/ScannerConfigBuilder.java
CodanBuilder
MissingBuilder
CBuilder
	AutotoolsBuildConfiguration
	CMakeBuildConfiguration
	MesonBuildConfiguration
	StandardBuildConfiguration
	ErrorBuildConfiguration
```
What do you think?

**Reply by jantje** [reply]
Thinking about this I think this is actually the only way to do this (unless the eclipse IDE would change which is highly unlikely)
Projects that do not support configurations will ignore the args for configuration and should build fine.
Projects that do support configurations and ignore the args will (probably) build the active configuration. Which can be seen as a bug in this project.
But in general it is the best option today with a possibility to ever have it better (if anyone ever cares)
Thanks 
Jantje

**Comment by jantje** [comment]
It looks like there exists a request for this https://bugs.eclipse.org/bugs/show_bug.cgi?id=313927.
I have my doubts about the proposed solution. (If I would not have doubts about the proposed solution I would not be me 😄 )

**Comment by jld01** [comment]
There is an added complication in that correct execution of the _CDT Builder_ for a particular build configuration may be dependent on prior execution of another (non-CDT) builder specified in the ordered list of configured builders. For example, the preceding builder may generate a C source file.
